chainerx_cc/chainerx/routines/math.cc

```diff
@@ -477,6 +477,50 @@ Array Sum(const Array& a, const OptionalAxes& axis, bool keepdims) {
     return out;
 }
 
+Array Prod(const Array& a, const OptionalAxes& axis, bool keepdims) {
+    Axes sorted_axis = internal::GetSortedAxesOrAll(axis, a.ndim());
+
+    // Decide the output dtype for integral input dtype.
+    Dtype out_dtype{};
+    switch (GetKind(a.dtype())) {
+        case DtypeKind::kBool:
+        case DtypeKind::kInt:  // fallthrough
+            out_dtype = Dtype::kInt64;
+            break;
+        case DtypeKind::kUInt:
+            out_dtype = Dtype::kInt64;  // TODO(niboshi): This should be kUInt64
+            break;
+        default:
+            out_dtype = a.dtype();
+    }
+
+    Array out = internal::EmptyReduced(a.shape(), out_dtype, sorted_axis, keepdims, a.device());
+    {
+        NoBackpropModeScope scope{};
+        a.device().backend().CallKernel<ProdKernel>(a, sorted_axis, out);
+    }
+
+    BackwardBuilder bb{"prod", a, out};
+    if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+        bt.Define([sorted_axis, a, out, in_shape = a.shape(), keepdims](BackwardContext& bctx) {
```

Comment:
Not sure if it's relevant, but I think `a` and `out` should be captured using `bb.RetainInput(0)` and `bb.RetainOutput(0)` respectively?

Reply:
Done.